typed/helpers.go

```diff
@@ -142,80 +123,50 @@ func handleAtom(a schema.Atom, tr schema.TypeRef, ah atomHandler) ValidationErro
 		name = "named type: " + *tr.NamedType
 	}
 
-	return ah.errorf("schema error: invalid atom: %v", name)
-}
-
-func (ef errorFormatter) validateScalar(t *schema.Scalar, v *value.Value, prefix string) (errs ValidationErrors) {
-	if v == nil {
-		return nil
-	}
-	if v.Null {
-		return nil
-	}
-	switch *t {
-	case schema.Numeric:
-		if v.FloatValue == nil && v.IntValue == nil {
-			// TODO: should the schema separate int and float?
-			return ef.errorf("%vexpected numeric (int or float), got %v", prefix, v)
-		}
-	case schema.String:
-		if v.StringValue == nil {
-			return ef.errorf("%vexpected string, got %v", prefix, v)
-		}
-	case schema.Boolean:
-		if v.BooleanValue == nil {
-			return ef.errorf("%vexpected boolean, got %v", prefix, v)
-		}
-	}
-	return nil
+	return errorf("schema error: invalid atom: %v", name)
 }
 
 // Returns the list, or an error. Reminder: nil is a valid list and might be returned.
-func listValue(val value.Value) (*value.List, error) {
-	switch {
-	case val.Null:
+func listValue(val value.Value) (value.List, error) {
+	if val.IsNull() {
 		// Null is a valid list.
 		return nil, nil
-	case val.ListValue != nil:
-		return val.ListValue, nil
-	default:
+	}
+	if !val.IsList() {
 		return nil, fmt.Errorf("expected list, got %v", val)
 	}
+	return val.List(), nil
 }
 
 // Returns the map, or an error. Reminder: nil is a valid map and might be returned.
-func mapValue(val value.Value) (*value.Map, error) {
-	switch {
-	case val.Null:
+func mapValue(val value.Value) (value.Map, error) {
+	if val == nil || val.IsNull() {
+		// Null is a valid map.
 		return nil, nil
-	case val.MapValue != nil:
-		return val.MapValue, nil
-	default:
+	}
+	if !val.IsMap() {
 		return nil, fmt.Errorf("expected map, got %v", val)
 	}
+	return val.Map(), nil
 }
 
 func keyedAssociativeListItemToPathElement(list *schema.List, index int, child value.Value) (fieldpath.PathElement, error) {
 	pe := fieldpath.PathElement{}
-	if child.Null {
+	if child.IsNull() {
 		// For now, the keys are required which means that null entries
 		// are illegal.
 		return pe, errors.New("associative list with keys may not have a null element")
 	}
-	if child.MapValue == nil {
+	if !child.IsMap() {
 		return pe, errors.New("associative list with keys may not have non-map elements")
 	}
 	keyMap := value.FieldList{}
```

Comment:
pre-sizing this slice might save an alloc or two. Also if we do any of the things we were talking about to value.Value, we could consider caching this once it's computed? (no action needed, just ideas for followup)